### ISSUE TITLE
Waive of "negative" test results

### ISSUE THREAD

**Issue by clebergnu** [avocado-framework/avocado#5227]
**Is your feature request related to a problem? Please describe.**
Recurring test jobs will eventually hit situations where the feature under test reveals itself straight out and consistently broken or "flaky", even without code changes that would be caught by pre-merge CIs.

For instance, on my development workstation, after the software under test receives an update (let's say an RPM package update), a test starts to fail because the update introduce some incompatibility or a bug.  Example:

```
$ avocado run integration.py 
JOB ID     : 1d8b28a04a5fb29bf72976cec851701b5a13f360
JOB LOG    : /home/cleber/avocado/job-results/job-2022-01-17T12.42-1d8b28a/job.log
 (2/2) integration.py:Subsystem.test_feature_bar: STARTED
 (1/2) integration.py:Subsystem.test_feature_foo: STARTED
 (2/2) integration.py:Subsystem.test_feature_bar: PASS (0.01 s)
 (1/2) integration.py:Subsystem.test_feature_foo: PASS (0.01 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-01-17T12.42-1d8b28a/results.html
JOB TIME   : 2.85 s

$ dnf update $package-being-tested-by-integration.py

$ avocado run integration.py 
JOB ID     : 430ddbadf987f29cfec2fb6282b5d3b8a40ccd3d
JOB LOG    : /home/cleber/avocado/job-results/job-2022-01-17T12.43-430ddba/job.log
 (1/2) integration.py:Subsystem.test_feature_foo: STARTED
 (2/2) integration.py:Subsystem.test_feature_bar: STARTED
 (1/2) integration.py:Subsystem.test_feature_foo: PASS (0.01 s)
 (2/2) integration.py:Subsystem.test_feature_bar: ERROR: None (0.03 s)
RESULTS    : PASS 1 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-01-17T12.43-430ddba/results.html
JOB TIME   : 2.96 s

$ echo $?
1
```

At this point, there are some options to address the problem, including:

1. Revert the package update
2. Disable the test
3. Immediately fix the package's bug
4. Immediately fix the test (if it turns out to be a test misconception about the tested software behavior)

But there are a few problems with those options:

- Reverting the update will only postpone the need for an actual resolution
- Disabling the test will make future job executions "blind" about fixes in either the test or software under test
- If the problem can not be addressed *immediately*, subsequent test executions will cause job failures

**Describe the solution you'd like**
A way of determining that a test's "negative" result (say a `FAIL` or `ERROR` status) should be ignored "for the time being".  For instance, if the `integration.py:Subsystem.test_feature_bar` is known to produce `ERROR` until a new package update is available with a fixed bug, one could:

```
$ avocado run --waive-results=integration.py:Subsystem.test_feature_bar --waive-as=WARN integration.py 
JOB ID     : 5470885c9fa9f22564ed59944bd199713f389abf
JOB LOG    : /home/cleber/avocado/job-results/job-2022-01-17T12.50-5470885/job.log
 (1/2) integration.py:Subsystem.test_feature_foo: STARTED
 (2/2) integration.py:Subsystem.test_feature_bar: STARTED
 (1/2) integration.py:Subsystem.test_feature_foo: PASS (0.01 s)
 (2/2) integration.py:Subsystem.test_feature_bar: WARN: ERROR waived (0.03 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 1 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-01-17T12.50-5470885/results.html
JOB TIME   : 2.92 s
```

The source of the list of tests that should be waived is yet to be defined, and it can include:

* A list based avocado setting entry
* A custom file format (say a `JSON` or `YAML` based)
* Tests marked with a given a given Avocado tag (say `:avocado: waive:FAIL`)

**Describe alternatives you've considered**
A less decoupled option would be a secondary command to be used in conjunction with a "vanilla" `avocado run`.  Example:

```
avocado run integration.py || avocado waive --waive-results=integration.py:Subsystem.test_feature_bar --waive-as=WARN latest
JOB ID     : 430ddbadf987f29cfec2fb6282b5d3b8a40ccd3d
JOB LOG    : /home/cleber/avocado/job-results/job-2022-01-17T12.43-430ddba/job.log
 (1/2) integration.py:Subsystem.test_feature_foo: STARTED
 (2/2) integration.py:Subsystem.test_feature_bar: STARTED
 (1/2) integration.py:Subsystem.test_feature_foo: PASS (0.01 s)
 (2/2) integration.py:Subsystem.test_feature_bar: ERROR: None (0.03 s)
RESULTS    : PASS 1 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-01-17T12.43-430ddba/results.html
JOB TIME   : 2.96 s

WAIVING OF JOB "latest" (/home/cleber/avocado/job-results/job-2022-01-17T12.43-430ddba)
 (2/2) integration.py:Subsystem.test_feature_bar: WARN: Waived ERROR (0.03 s)

$ echo $?
0
```

**Comment by beraldoleal** [comment]
Hi @clebergnu thanks for that. A few questions:

> The source of the list of tests that should be waived is yet to be defined, and it can include:
> 
>     * A list based avocado setting entry
> 
>     * A custom file format (say a `JSON` or `YAML` based)
> 
>     * Tests marked with a given a given Avocado tag (say `:avocado: waive:FAIL`)

IIUC, would be nice to add this support to not only python code, but other types. Because of that my only concern here is to implement *only* the last one. Are you talking about implementing all options above or should we choose one?

Also how should we flag the other output formats? tap, xunit? I mean, from the CI perspective, is the Avocado responsibility to let CI aware of this wave? (I haven't investigated if any of those formats has support to something like this).

I guess what I'm missing is how this wave will be respected at the CI level (when reporting tests individually)? I know that Gitlab, it was using the "quarentine" logic internally, but I'm not sure if this is available for any user via their runners.

In any case, one of my concerns right now is the consistency between the Human Output format and others. Or how a tap or xUnit will present that information for consistency.

**Comment by beraldoleal** [comment]
Just sharing what I found so far:

> * Tests marked with a given a given Avocado tag (say `:avocado: waive:FAIL`)

unittest has support to "[XFAIL](https://docs.python.org/3/library/unittest.html#unittest.expectedFailure)", I'm not sure if it is 100% what we are looking for, but maybe we should consider it as well.

Some runners with support to this feature, is using it as well on xunit output:

```
$ python -m xmlrunner test.py
$ cat TEST-test.TestStringMethods-20220118102921.xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="test.TestStringMethods-20220118102921" tests="3" file="test.py" time="0.004" timestamp="2022-01-18T10:29:21" failures="0" errors="0" skipped="1">
	<testcase classname="test.TestStringMethods" name="test_isupper" time="0.004" timestamp="2022-01-18T10:29:21" file="test.py" line="8"/>
	<testcase classname="test.TestStringMethods" name="test_upper" time="0.000" timestamp="2022-01-18T10:29:21" file="test.py" line="5"/>
	<testcase classname="test.TestStringMethods" name="test_split" time="0.000" timestamp="2022-01-18T10:29:21" file="test.py" line="12">
		<skipped type="XFAIL" message="expected failure: (&lt;class 'AssertionError'&gt;, AssertionError('Should fail because issue #32'), &lt;traceback object at 0x7fce97200340&gt;)"/>
	</testcase>
</testsuite>
```

*If this is a valid option*, then we just need to think on how to map the list of "waved" configured by the user with this output. What I liked about this approach:

 1. Perhaps CI in the future will show the "type" field as well (I'm not sure if any is supporting already);
 2. CI is not breaking because the test is flagged as skipped;
 3. Third-party tools could parse this script to get the "XFAIL" type reason.

**Comment by clebergnu** [comment]
> Hi @clebergnu thanks for that. A few questions:
> 
> > The source of the list of tests that should be waived is yet to be defined, and it can include:
> > ```
> > * A list based avocado setting entry
> > 
> > * A custom file format (say a `JSON` or `YAML` based)
> > 
> > * Tests marked with a given a given Avocado tag (say `:avocado: waive:FAIL`)
> > ```
> 
> IIUC, would be nice to add this support to not only python code, but other types. Because of that my only concern here is to implement _only_ the last one. Are you talking about implementing all options above or should we choose one?
> 

The different examples cover different use cases, so I believe we'll need more than one.

> Also how should we flag the other output formats? tap, xunit? I mean, from the CI perspective, is the Avocado responsibility to let CI aware of this wave? (I haven't investigated if any of those formats has support to something like this).
> 

Of course.  The example given uses the Human UI output, but we should add the best mapping possible for the "waive" condition/status for all other result formats.  About "avocado will be responsible for letting the CI aware of the waive", it's certainly aligned with what we need, but a bit of a stretch because of the (current) limited communication from Avocado and CI environments.  The direct coupling and responsibility at this point is with the test result formats, which in turn, should be useful enough for CI environments capable of using them.

> I guess what I'm missing is how this wave will be respected at the CI level (when reporting tests individually)? I know that Gitlab, it was using the "quarentine" logic internally, but I'm not sure if this is available for any user via their runners.
> 

If the waive is handled at the Avocado job level, and communicated to the test result formats that CIs such as GitLab can use, then I believe we're all set.

> In any case, one of my concerns right now is the consistency between the Human Output format and others. Or how a tap or xUnit will present that information for consistency.

Sure, I wasn't explicit in the example, but by adding a new test status logic, it should be mapped to all test result formats (xUnit, TAP, etc).

**Comment by clebergnu** [comment]
> Just sharing what I found so far:
> 
> > * Tests marked with a given a given Avocado tag (say `:avocado: waive:FAIL`)
> 
> unittest has support to "[XFAIL](https://docs.python.org/3/library/unittest.html#unittest.expectedFailure)", I'm not sure if it is 100% what we are looking for, but maybe we should consider it as well.
> 
> Some runners with support to this feature, is using it as well on xunit output:
> 
> ```
> $ python -m xmlrunner test.py
> $ cat TEST-test.TestStringMethods-20220118102921.xml
> <?xml version="1.0" encoding="UTF-8"?>
> <testsuite name="test.TestStringMethods-20220118102921" tests="3" file="test.py" time="0.004" timestamp="2022-01-18T10:29:21" failures="0" errors="0" skipped="1">
> 	<testcase classname="test.TestStringMethods" name="test_isupper" time="0.004" timestamp="2022-01-18T10:29:21" file="test.py" line="8"/>
> 	<testcase classname="test.TestStringMethods" name="test_upper" time="0.000" timestamp="2022-01-18T10:29:21" file="test.py" line="5"/>
> 	<testcase classname="test.TestStringMethods" name="test_split" time="0.000" timestamp="2022-01-18T10:29:21" file="test.py" line="12">
> 		<skipped type="XFAIL" message="expected failure: (&lt;class 'AssertionError'&gt;, AssertionError('Should fail because issue #32'), &lt;traceback object at 0x7fce97200340&gt;)"/>
> 	</testcase>
> </testsuite>
> ```
> 

`XFAIL` is probably the closest test status I know by now, but it's still significant different than a waived test result.  For instance, a truth table for a waived and a "expected to fail" test is:

```
+------------+-----------+--------+
|Test Status | Operation | Result |
+------------+-----------+--------+
| FAIL       | WAIVE     | PASS   |
| PASS       | WAIVE     | PASS   |
| FAIL       | XFAIL     | PASS   |
| PASS       | XFAIL     | FAIL   |
+------------+-----------+--------+
```

The "If the test passes, it will be considered a failure" behavior is not what we want.  

> _If this is a valid option_, then we just need to think on how to map the list of "waved" configured by the user with this output. What I liked about this approach:
> 
> 1. Perhaps CI in the future will show the "type" field as well (I'm not sure if any is supporting already);

Indeed, this is something we should pursue or check if it exists, in parallel to solving the issue at the Avocado runner level.  Checking with say, GitLab CI, by searching for the feature or issues (or opening new ones) is a relevant task for this issue.  But, as stated before, I don't think an "expected failure" status is exactly what we want here, so this could mean investigating adding this new type of status to those test results (and consequently to the CI results).

> 2. CI is not breaking because the test is flagged as skipped;

Yes, I hope no runner will signal a non-zero status code for the test job, which is what CI environments primarily consider as the overall PASS/FAIL status.  The data on a "xUnit" kind of result is secondary to that AFAICT.

> 3. Third-party tools could parse this script to get the "XFAIL" type reason.

Do you mean that third-party tools could parse such "xUnit" result file and get the "reason" from, say, the "message" attribute?

**Comment by beraldoleal** [comment]
> The different examples cover different use cases, so I believe we'll need more than one.

Could you please elaborate on the test cases, please? IIUC, we need to inform a test is "waived" and this should be valid for any type of test. It sounds strange to me to waive differently deepening on the test type.

> The "If the test passes, it will be considered a failure" behavior is not what we want.

Yes, indeed. This was something that caught my attention too. And that is the reason that I spotted the "I'm not sure if it is 100% what we are looking for" ;)

> Yes, I hope no runner will signal a non-zero status code for the test job, which is what CI environments primarily consider as the overall PASS/FAIL status. The data on a "xUnit" kind of result is secondary to that AFAICT.

Right, but even if secondary, we just need to keep in mind the consistency.  A job passing (exit code = 0) and a test report with any ERROR would be at least confusing. However by looking at the desired "truth table", this will not happen.

However, another question would be: Wouldn't the "PASS" status on a "WAIVED FAILED", give the impression that everything is ok?  Thinking about someone that was not aware of the WAVE and it is looking only at the CI results. Probably, users tend to not look at the log output if a "test report" is available.

Developers might forget about it, since the waive is not displayed on CIs results (and probably there is no much we can by now).

Besides that, it is just a matter of *also* setting the "type" properly the xunit file. To something like: WAVED.

> Do you mean that third-party tools could parse such "xUnit" result file and get the "reason" from, say, the "message" attribute?

The "type" and "message" attributes are available on xUnit result, indeed. But even better, gitlabCI for instance it will paste as "test output" those fields on the test report. So third-party apps will not need to get the entire result and scan for the test result again.
